File: sdk/python/memg/context.py
"""Context builder: merges conscious, recalled, and summary facts into a prompt string."""

from __future__ import annotations

import math
import re
import unicodedata
from typing import List, Optional

from .types import Artifact, ConsciousFact, RecalledFact, RecalledSummary, TurnSummary

_DEDUP_PREFIXES = ("the user ", "user ", "user's ", "[historical] ")


def _normalize_for_dedup(s: str) -> str:
    s = s.lower().strip()
    for prefix in _DEDUP_PREFIXES:
        if s.startswith(prefix):
            s = s[len(prefix):]
    return s.strip()


def _estimate_tokens(s: str) -> int:
    words = len(s.split())
    if words > 0:
        return math.ceil(words * 1.3)
    # No spaces (likely CJK) — count Unicode code points, not bytes.
    runes = sum(1 for _ in s)
    if runes == 0:
        return 0
    return math.ceil(runes * 2 / 3)


def build_context(
    conscious_facts: Optional[List[ConsciousFact]] = None,
    recalled_facts: Optional[List[RecalledFact]] = None,
    summaries: Optional[List[RecalledSummary]] = None,
    turn_summaries: Optional[List[TurnSummary]] = None,
    artifacts: Optional[List[Artifact]] = None,
    total_token_budget: int = 4000,
    summary_token_budget: int = 1000,
) -> str:
    """Build a context string from memory components.

    Priority: conscious > turn_summaries > recalled > artifacts > summaries.
    Deduplicates by normalized content across components.
    """
    budget = total_token_budget if total_token_budget > 0 else 4000
    summary_budget = summary_token_budget if summary_token_budget > 0 else 1000

    parts: list = []
    seen: set = set()
    tokens_used = 0

    # 1. Conscious facts (highest priority)
    if conscious_facts:
        section_lines = ["User profile:"]
        for f in conscious_facts:
            normalized = _normalize_for_dedup(f.content)
            if normalized in seen:
                continue
            seen.add(normalized)
            line = f"- {f.content}"
            est = _estimate_tokens(line + "\n")
            if tokens_used + est > budget:
                break
            section_lines.append(line)
            tokens_used += est
        if len(section_lines) > 1:
            parts.append("\n".join(section_lines))

    # 2. Session context (turn summaries)
    if turn_summaries:
        section_lines = ["\nSession context:"]
        header_tokens = _estimate_tokens(section_lines[0])
        section_tokens = header_tokens
        any_added = False

        for ts in turn_summaries:
            if ts.is_overview:
                line = f"- [Overview] {ts.summary}"
            else:
                line = f"- [Turns {ts.start_turn}-{ts.end_turn}] {ts.summary}"
            est = _estimate_tokens(line)
            if tokens_used + section_tokens + est > budget:
                break
            section_lines.append(line)
            section_tokens += est
            any_added = True

        if any_added:
            parts.append("\n".join(section_lines))
            tokens_used += section_tokens

    # 3. Recalled facts (medium priority) - dedup against conscious
    if recalled_facts:
        section_lines = ["\nRelevant context from memory:"]
        header_tokens = _estimate_tokens(section_lines[0])
        section_tokens = header_tokens
        any_added = False

        for f in recalled_facts:
            normalized = _normalize_for_dedup(f.content)
            if normalized in seen:
                continue
            seen.add(normalized)
            prefix = "- "
            if f.temporal_status == "historical":
                prefix = "- [historical] "
            line = prefix + f.content
            est = _estimate_tokens(line)
            if tokens_used + section_tokens + est > budget:
                break
            section_lines.append(line)
            section_tokens += est
            any_added = True

        if any_added:
            parts.append("\n".join(section_lines))
            tokens_used += section_tokens

    # 4. Relevant code/data (artifacts)
    if artifacts:
        section_lines = ["\nRelevant code/data:"]
        header_tokens = _estimate_tokens(section_lines[0])
        section_tokens = header_tokens
        any_added = False

        for a in artifacts:
            entry_parts = []
            label = f"[{a.artifact_type}"
            if a.language:
                label += f":{a.language}"
            label += f"] {a.description}"
            entry_parts.append(f"- {label}")
            if a.artifact_type == "code" and a.language:
                entry_parts.append(f"  ```{a.language}")
                entry_parts.append(f"  {a.content}")
                entry_parts.append("  ```")
            else:
                entry_parts.append(f"  {a.content}")
            entry = "\n".join(entry_parts)
            est = _estimate_tokens(entry)
            if tokens_used + section_tokens + est > budget:
                break
            section_lines.append(entry)
            section_tokens += est
            any_added = True

        if any_added:
            parts.append("\n".join(section_lines))
            tokens_used += section_tokens

    # 5. Summaries (lowest priority, own sub-budget)
    if summaries:
        effective_budget = summary_budget
        remaining = budget - tokens_used
        if effective_budget > remaining:
            effective_budget = remaining

        if effective_budget > 0:
            section_lines = ["\nRelevant past conversations:"]
            header_tokens = _estimate_tokens(section_lines[0])
            section_tokens = header_tokens
            any_added = False

            for s in summaries:
                normalized = _normalize_for_dedup(s.summary)
                if normalized in seen:
                    continue
                seen.add(normalized)
                date_str = ""
                if s.created_at:
                    date_str = s.created_at.strftime("%b {}, %Y").format(s.created_at.day)
                line = f"- [{date_str}] {s.summary}" if date_str else f"- {s.summary}"
                est = _estimate_tokens(line)
                if section_tokens + est > effective_budget:
                    break
                section_lines.append(line)
                section_tokens += est
                any_added = True

            if any_added:
                parts.append("\n".join(section_lines))

    result = "\n".join(parts)
    return result.rstrip("\n")


def adaptive_window_size(total_budget: int, used_tokens: int, default_turns: int) -> int:
    """Calculate how many recent messages to load based on remaining token budget.

    Args:
        total_budget: Overall token budget.
        used_tokens: Tokens already consumed by other context components.
        default_turns: Default/maximum number of turns.

    Returns:
        Number of messages to include in the working memory window.
    """
    remaining = total_budget - used_tokens
    floor_tokens = 800
    if remaining < floor_tokens:
        remaining = floor_tokens
    avg_tokens_per_msg = 200
    max_msgs = remaining // avg_tokens_per_msg
    if max_msgs > default_turns:
        max_msgs = default_turns
    min_msgs = floor_tokens // avg_tokens_per_msg
    if max_msgs < min_msgs:
        max_msgs = min_msgs
    return max_msgs

File: sdk/typescript/src/context.ts
/**
 * Context builder: merges conscious facts, recalled facts, and summaries
 * into a single context string under a token budget. Matches Go's BuildContext.
 */

import type { Artifact, ConsciousFact, RecalledFact, RecalledSummary, TurnSummary } from './types.js';

export interface ContextBudget {
  totalTokens: number;
  summaryTokens: number;
}

export interface ContextInput {
  consciousFacts: ConsciousFact[];
  recalledFacts: RecalledFact[];
  summaries: RecalledSummary[];
  turnSummaries?: TurnSummary[];
  artifacts?: Artifact[];
  budget: ContextBudget;
}

/**
 * Token estimation: ceil(wordCount * 1.3).
 * For CJK text with few spaces, falls back to rune-based estimation.
 */
export function estimateTokens(s: string): number {
  const words = s.split(/\s+/).filter(Boolean).length;
  if (words > 0) {
    return Math.ceil(words * 1.3);
  }
  // No spaces (likely CJK) — roughly 1 token per 1.5 characters.
  const runes = [...s].length;
  if (runes === 0) return 0;
  return Math.ceil((runes * 2) / 3);
}

/**
 * Normalize content for dedup. Strips common prefixes, lowercases, trims.
 */
function normalizeForDedup(s: string): string {
  let n = s.toLowerCase().trim();
  for (const prefix of ['the user ', 'user ', "user's ", '[historical] ']) {
    if (n.startsWith(prefix)) {
      n = n.slice(prefix.length);
    }
  }
  return n.trim();
}

/**
 * Format a date string as "Jan 2, 2006" for summary display.
 */
function formatSummaryDate(isoDate: string | undefined): string {
  if (!isoDate) return 'Unknown date';
  try {
    const d = new Date(isoDate);
    const months = [
      'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
      'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec',
    ];
    return `${months[d.getMonth()]} ${d.getDate()}, ${d.getFullYear()}`;
  } catch {
    return 'Unknown date';
  }
}

/**
 * Build context string. Priority: conscious > turn summaries > recalled > artifacts > summaries.
 * Dedup by normalized content across all components.
 */
export function buildContext(input: ContextInput): string {
  let budget = input.budget.totalTokens;
  if (budget <= 0) budget = 4000;
  let summaryBudget = input.budget.summaryTokens;
  if (summaryBudget <= 0) summaryBudget = 1000;

  const parts: string[] = [];
  const seen = new Set<string>();
  let tokensUsed = 0;

  // 1. Conscious facts (highest priority).
  if (input.consciousFacts.length > 0) {
    const lines: string[] = ['User profile:'];
    for (const f of input.consciousFacts) {
      const normalized = normalizeForDedup(f.content);
      if (seen.has(normalized)) continue;
      seen.add(normalized);
      const line = `- ${f.content}`;
      const est = estimateTokens(line + '\n');
      if (tokensUsed + est > budget) break;
      lines.push(line);
      tokensUsed += est;
    }
    if (lines.length > 1) {
      parts.push(lines.join('\n'));
    }
  }

  // 2. Session context (turn summaries).
  const turnSummaries = input.turnSummaries ?? [];
  if (turnSummaries.length > 0) {
    const header = '\nSession context:';
    const headerTokens = estimateTokens(header + '\n');
    let sectionTokens = headerTokens;
    const lines: string[] = [header];
    let any = false;

    for (const ts of turnSummaries) {
      let line: string;
      if (ts.isOverview) {
        line = `- [Overview] ${ts.summary}`;
      } else {
        line = `- [Turns ${ts.startTurn}-${ts.endTurn}] ${ts.summary}`;
      }
      const est = estimateTokens(line + '\n');
      if (tokensUsed + sectionTokens + est > budget) break;
      lines.push(line);
      sectionTokens += est;
      any = true;
    }

    if (any) {
      parts.push(lines.join('\n'));
      tokensUsed += sectionTokens;
    }
  }

  // 3. Recalled facts (medium priority) — dedup against conscious.
  if (input.recalledFacts.length > 0) {
    const header = '\nRelevant context from memory:';
    const headerTokens = estimateTokens(header + '\n');
    let sectionTokens = headerTokens;
    const lines: string[] = [header];
    let any = false;

    for (const f of input.recalledFacts) {
      const normalized = normalizeForDedup(f.content);
      if (seen.has(normalized)) continue;
      seen.add(normalized);
      let line = '- ';
      if (f.temporalStatus === 'historical') {
        line += '[historical] ';
      }
      line += f.content;
      const est = estimateTokens(line + '\n');
      if (tokensUsed + sectionTokens + est > budget) break;
      lines.push(line);
      sectionTokens += est;
      any = true;
    }

    if (any) {
      parts.push(lines.join('\n'));
      tokensUsed += sectionTokens;
    }
  }

  // 4. Relevant code/data (artifacts).
  const artifacts = input.artifacts ?? [];
  if (artifacts.length > 0) {
    const header = '\nRelevant code/data:';
    const headerTokens = estimateTokens(header + '\n');
    let sectionTokens = headerTokens;
    const lines: string[] = [header];
    let any = false;

    for (const a of artifacts) {
      let entry = `- [${a.artifactType}`;
      if (a.language) {
        entry += `:${a.language}`;
      }
      entry += `] ${a.description}\n`;
      if (a.artifactType === 'code' && a.language) {
        entry += `  \`\`\`${a.language}\n`;
        entry += `  ${a.content}\n`;
        entry += '  ```\n';
      } else {
        entry += `  ${a.content}\n`;
      }

      const est = estimateTokens(entry);
      if (tokensUsed + sectionTokens + est > budget) break;
      lines.push(entry.replace(/\n$/, ''));
      sectionTokens += est;
      any = true;
    }

    if (any) {
      parts.push(lines.join('\n'));
      tokensUsed += sectionTokens;
    }
  }

  // 5. Summaries (lowest priority, own sub-budget).
  if (input.summaries.length > 0) {
    let effectiveBudget = summaryBudget;
    const remaining = budget - tokensUsed;
    if (effectiveBudget > remaining) effectiveBudget = remaining;

    if (effectiveBudget > 0) {
      const header = '\nRelevant past conversations:';
      const headerTokens = estimateTokens(header + '\n');
      let sectionTokens = headerTokens;
      const lines: string[] = [header];
      let any = false;

      for (const s of input.summaries) {
        const normalized = normalizeForDedup(s.summary);
        if (seen.has(normalized)) continue;
        seen.add(normalized);
        const dateStr = formatSummaryDate(s.createdAt);
        const line = `- [${dateStr}] ${s.summary}`;
        const est = estimateTokens(line + '\n');
        if (sectionTokens + est > effectiveBudget) break;
        lines.push(line);
        sectionTokens += est;
        any = true;
      }

      if (any) {
        parts.push(lines.join('\n'));
      }
    }
  }

  return parts.join('\n').replace(/\n+$/, '');
}

/**
 * Calculate how many recent messages to load based on remaining token budget.
 * Matches Go AdaptiveWindowSize.
 */
export function adaptiveWindowSize(
  totalBudget: number,
  usedTokens: number,
  defaultTurns: number
): number {
  let remaining = totalBudget - usedTokens;
  const floorTokens = 800;
  if (remaining < floorTokens) remaining = floorTokens;
  const avgTokensPerMsg = 200;
  let maxMsgs = Math.floor(remaining / avgTokensPerMsg);
  if (maxMsgs > defaultTurns) maxMsgs = defaultTurns;
  const minMsgs = Math.floor(floorTokens / avgTokensPerMsg);
  if (maxMsgs < minMsgs) maxMsgs = minMsgs;
  return maxMsgs;
}
